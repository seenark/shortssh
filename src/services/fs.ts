import * as fs from "node:fs/promises"
import { ErrorHelpers } from "@/helpers"
import { Data, Effect } from "effect"

export class ReadDirError extends Data.TaggedError("fs/ReadDirError")<ErrorHelpers.ErrorMsg> {
  static new = ErrorHelpers.createErrorFactory(this)
}

export function readDir(dir: string) {
  return Effect.tryPromise({
    catch: ReadDirError.new(),
    try: () => fs.readdir(dir, { withFileTypes: true }),
  })
}

export class FsContext extends Effect.Service<FsContext>()("services/Fs", {
  succeed: {
    readDir,
  },
}) { }
