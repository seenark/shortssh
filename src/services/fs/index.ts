import * as fs from "node:fs/promises"
import { join } from "node:path"
import { ErrorHelpers } from "@/helpers"
import { Array, Data, Effect } from "effect"

class ReadDirError extends Data.TaggedError("fs/ReadDirError")<ErrorHelpers.ErrorMsg> {
  static new = ErrorHelpers.createErrorFactory(this)
}

function readDir(dir: string) {
  return Effect.tryPromise({
    catch: ReadDirError.new(),
    try: () => fs.readdir(dir, { withFileTypes: true }),
  })
}

function scanSshConfig(dir: string): Effect.Effect<string[], ReadDirError> {
  return readDir(dir).pipe(
    Effect.flatMap(Effect.forEach((d) => {
      const fulllpath = join(dir, d.name)
      if (d.isDirectory())
        return scanSshConfig(fulllpath)
      if (d.name === "config")
        return Effect.succeed([fulllpath])
      return Effect.succeed([])
    }, { concurrency: "unbounded" })),
    Effect.map(Array.flatten),
    Effect.map(d => d),
  )
}

export function findSshConfigs(baseDir: string) {
  return scanSshConfig(baseDir)
}
