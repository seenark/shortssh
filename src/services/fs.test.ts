import type { Dirent } from "node:fs"
import fs from "node:fs/promises"
import { Effect } from "effect"
import { beforeEach, describe, expect, it, vi } from "vitest"
import { readDir } from "./fs"

vi.mock("node:fs/promises")

describe("fs service", () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  it("should successfully read directory contents", async () => {
    // Arrange
    const mockDirents = [
      { isFile: () => true, name: "file1.txt" } as Dirent,
      { isDirectory: () => true, name: "dir1" } as Dirent,
    ]
    vi.mocked(fs.readdir).mockResolvedValue(mockDirents)
    const testDir = "/test/dir"

    // Act
    const result = await Effect.runPromise(readDir(testDir))

    // Assert
    expect(result).toEqual(mockDirents)
    expect(fs.readdir).toHaveBeenCalledWith(testDir, { withFileTypes: true })
    expect(fs.readdir).toHaveBeenCalledTimes(1)
  })

  it("should return ReadDirError when fs.readdir fails", async () => {
    // Arrange
    const testDir = "/test/dir"
    const fsError = new Error("Failed to read directory")
    vi.mocked(fs.readdir).mockRejectedValue(fsError)

    // Act
    const program = readDir(testDir)

    // Assert
    await expect(Effect.runPromise(program)).rejects.toMatchSnapshot()
    expect(fs.readdir).toHaveBeenCalledWith(testDir, { withFileTypes: true })
    expect(fs.readdir).toHaveBeenCalledTimes(1)
  })
})
