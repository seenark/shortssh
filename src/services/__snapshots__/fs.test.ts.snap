// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fs service > should return ReadDirError when fs.readdir fails 1`] = `[(FiberFailure) fs/ReadDirError: An error has occurred]`;
